### PR TITLE
Add preprocessing functions for Llama4 multimodal (tiles)

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description

Image preprocessing utils for Llama4 multimodal. Core workflow in `pre_process_llama4_image()`. Input example: jnp.array of (536, 640, 3); Output example: (5, 3, 336, 336).
- Find best Resolution Canvas: `get_best_resolution` determines the best canvas resolution for the input image from a list of supported resolutions.
- Padding: `pad_to_best_fit_jax` pads the image to fit the determined best resolution canvas.
- Normalization: Normalizes the pixel values of the padded image.
- Tiling: `split_to_tiles_jax` Splits the processed image into a number of tiles.
- Global Tile Addition: If the image is split into multiple tiles, an additional global tile (a resized/normalized version of the original image) is added to the set of tiles.

# Tests

Passing unit tests, processed images match with Llama4 from HuggingFace transformers.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/tests/pyconfig_test.py
Comment (hengtaoguo) [inline MaxText/tests/pyconfig_test.py:1]: This change was generated from `bash code_style.sh`

**Review by ZhaoyueCheng — APPROVED** [review]

**Review by aireenmei — APPROVED** [review]
LGTM, thanks Hengtao!

File: MaxText/multimodal_utils.py [hunk 1]
```diff
@@ -42,6 +43,13 @@
 # One MEDIA means one image or multiple images in one video, but now we only support one image
 GEMMA_NUM_TOKENS_PER_MEDIA = GEMMA_NUM_PLACEHOLDER_TOKENS_PER_IMAGE + 4
 
+# Constants for Llama4-specific processing
```
Comment (gobbleturk) [inline MaxText/multimodal_utils.py:46]: perhaps a single file containing all constants (this is fine for now with only 2 models)

File: MaxText/multimodal_utils.py [hunk 2]
```diff
@@ -70,6 +78,190 @@ def _normalize_images(images, mean, std):
   return images
 
 
+def get_factors(dividend: int):
```
Comment (gobbleturk) [inline MaxText/multimodal_utils.py:81]: I think we can just use block_size * (x // block_size + 1) (rounded up)

File: MaxText/tests/multimodal_utils_test.py [hunk 3]
```diff
@@ -64,5 +65,56 @@ def test_add_images_in_batch(self):
...
+class TestLlama4ImageProcessing(unittest.TestCase):
+  """Test Llama4 image processing"""
+
+  def setUp(self):
+    super().setUp()
+    self.LLAMA4_TILES_NUM = 16
+    self.LLAMA4_TILE_SIZE = 336
+    self.NUM_IMAGE_CHANNELS = 3
+
+  def test_get_best_resolution(self):
```
Comment (gobbleturk) [inline MaxText/tests/multimodal_utils_test.py:77]: possibly this can be removed with refactor

num_tiles_x_direction = math.ceil(x_size/x_block_size)

File: MaxText/tests/pyconfig_test.py [hunk 4]
```diff
@@ -120,7 +120,7 @@ def test_overriding_model(self):
         skip_jax_distributed_system=True,
         model_name="gemma-7b",
         override_model_config=True,
-        base_emb_dim=1024, # Defined as 3072 in gemma-7b
+        base_emb_dim=1024,  # Defined as 3072 in gemma-7b
```
Comment (gobbleturk) [inline MaxText/tests/pyconfig_test.py:123]: mb.

**Review by gobbleturk — APPROVED** [review]
Awesome!!!

File: MaxText/multimodal_utils.py [hunk 5]
```diff
@@ -70,6 +78,190 @@ def _normalize_images(images, mean, std):
...
+  Returns:
+      set: A set containing all factors of the number.
+  """
+  factors_set = set()
+
+  for i in range(1, int(dividend**0.5) + 1):
+    if dividend % i == 0:
+      factors_set.add(i)
+      factors_set.add(dividend // i)
+  return factors_set
```
Comment (gobbleturk) [inline MaxText/multimodal_utils.py:96]: possibly there is a math built in for this

**Review by gobbleturk — APPROVED** [review]

File: MaxText/multimodal_utils.py [hunk 6]
```diff
@@ -70,6 +78,190 @@ def _normalize_images(images, mean, std):
...
+      image_width (int): The width of the image.
+      possible_resolutions (list): A list of possible resolutions.
+      resize_to_max_canvas (bool): Whether to resize to max canvas or not.
+  Returns:
+      tuple: The best resolution for the image.
+  """
+  if resize_to_max_canvas:
+    return max(possible_resolutions, key=lambda x: x[0] * x[1])
+  else:
+    return min(possible_resolutions, key=lambda x: abs(x[0] - img_height) + abs(x[1] - image_width))
```
Comment (gobbleturk) [inline MaxText/multimodal_utils.py:132]: okay I started at this code a bit more - I don't think this is the same as ceil(x/block_x) - in some cases it will round down, e.g. if the image is 201 x 201 and the block size is 100 x 100 I think it will still choose to use 2x2 blocks? Some pixels are are not in these four blocks but at least they will appear in the global image

**Comment by hengtaoguo** [comment]
> Awesome!!!

Thank you so much for the thorough review!

File: MaxText/multimodal_utils.py (hunk 1 above)
Comment (hengtaoguo) [reply MaxText/multimodal_utils.py:47]: Great suggestion! I will refactor these variables to a separate file for multimodal stuff.

File: MaxText/multimodal_utils.py (hunk 6 above)
Comment (hengtaoguo) [reply MaxText/multimodal_utils.py:134]: Thank Matt for the great catch! This is line is to find the resolution closest to the original image dimensions (minimizing padding/cropping), it could be slightly smaller than the actual image size so the ceiling might not work here. Yeah let's keep it the current way to temporarily unblock downstream tasks!
